### ISSUE TITLE
enhanced memcached sent invalid trailer with nginx 1.7.4

### ISSUE THREAD

**Issue by manwe-pl** [bpaquet/ngx_http_enhanced_memcached_module#5]
Hi. With version 1.4.7 module is working fine, exact same config with 1.7.4 throws:

2014/08/12 20:49:30 [error] 2276#0: *57 enhanced memcached sent invalid trailer while reading response header from upstream, client: XX.XX.XX.XX, server: yyyyyyyyyy, request: "GET /test.php HTTP/1.1", upstream: "memcached://127.0.0.1:11211", host: "yyyyyyyyyy"

2014/08/12 20:49:30 [alert] 2276#0: *57 writev() failed (22: Invalid argument) while sending to client, client: XX.XX.XX.XX, server: yyyyyyyyyy, request: "GET /test.php HTTP/1.1", upstream: "memcached://127.0.0.1:11211", host: "yyyyyyyyyy"

Any ideas? :)

**Comment by tiberius** [comment]
Hi, 

seems like this problem has started popping up again using current master and recent nginx versions (tested 1.10.3 as well as 1.11.13):

```
2017/04/12 12:49:46 [error] 18#18: *21 enhanced memcached: sent invalid trailer while reading response header from upstream, client: 172.17.0.1, server: example.xxx, request: "GET / HTTP/1.1", upstream: "memcached://unix:/var/lock/memcached.sock:", host: "example.xxx"
```

nginx build config:
```
# nginx -V
nginx version: nginx/1.11.13
built by gcc 4.9.2 (Debian 4.9.2-10) 
built with OpenSSL 1.0.1t  3 May 2016
TLS SNI support enabled
configure arguments: --prefix=/etc/nginx --sbin-path=/usr/sbin/nginx --modules-path=/usr/lib/nginx/modules --conf-path=/etc/nginx/nginx.conf --error-log-path=/var/log/nginx/error.log --http-log-path=/var/log/nginx/access.log --pid-path=/var/run/nginx.pid --lock-path=/var/run/nginx.lock --http-client-body-temp-path=/var/cache/nginx/client_temp --http-proxy-temp-path=/var/cache/nginx/proxy_temp --http-fastcgi-temp-path=/var/cache/nginx/fastcgi_temp --http-uwsgi-temp-path=/var/cache/nginx/uwsgi_temp --http-scgi-temp-path=/var/cache/nginx/scgi_temp --user=nginx --group=nginx --with-compat --with-file-aio --with-threads --with-http_addition_module --with-http_auth_request_module --with-http_dav_module --with-http_flv_module --with-http_gunzip_module --with-http_gzip_static_module --with-http_mp4_module --with-http_random_index_module --with-http_realip_module --add-module=/usr/src/ngx_pagespeed-1.12.34.2-beta/ --add-module=/usr/src/ngx_http_enhanced_memcached_module --add-module=/usr/src/headers-more-nginx-module-0.32 --with-http_secure_link_module --with-http_slice_module --with-http_ssl_module --with-http_stub_status_module --with-http_sub_module --with-http_v2_module --with-mail --with-mail_ssl_module --with-stream --with-stream_realip_module --with-stream_ssl_module --with-stream_ssl_preread_module --with-cc-opt='-g -O2 -fstack-protector-strong -Wformat -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -fPIC' --with-ld-opt='-Wl,-z,relro -Wl,-z,now -Wl,--as-needed -pie'
```

Let me know if I you need something else.

**Comment by manwe-pl** [comment]
I've just tested 1.12.0 (1.11.13 as stable) and everything is working fine.

`# nginx -V
nginx version: nginx/1.12.0
built with OpenSSL 1.0.2k  26 Jan 2017
TLS SNI support enabled
configure arguments: --prefix=/var/lib/nginx --sbin-path=/usr/sbin/nginx --conf-path=/etc/nginx/nginx.conf --pid-path=/run/nginx.pid --lock-path=/run/lock/nginx.lock --with-cc-opt=-I/usr/include --with-ld-opt=-L/usr/lib64 --http-log-path=/var/log/nginx/access.log --error-log-path=/var/log/nginx/error.log --http-client-body-temp-path=tmp/client --http-proxy-temp-path=tmp/proxy --http-fastcgi-temp-path=tmp/fastcgi --http-scgi-temp-path=tmp/scgi --http-uwsgi-temp-path=tmp/uwsgi --modules-path=modules --with-pcre --with-threads --without-stream_geo_module --without-stream_split_clients_module --with-http_addition_module --with-http_auth_request_module --with-http_geoip_module --with-http_gunzip_module --with-http_gzip_static_module --with-http_realip_module --with-http_stub_status_module --with-http_sub_module --with-http_v2_module --add-module=ext/ngx_devel_kit-0.3.0 --add-module=ext/set-misc-nginx-module-0.31 --add-module=ext/echo-nginx-module-cca11509d2a0e31f54226f5fb5b1b522f9040106 --add-module=ext/lua-nginx-module-0.10.8 --add-module=ext/lua-upstream-nginx-module-0.06 --add-module=ext/ngx_http_enhanced_memcached_module-nginx_1.11.9 --add-module=ext/nginx_tcp_proxy_module-nginx_1.11.9 --add-module=ext/nginx-http-rdns-1.0 --add-module=ext/headers-more-nginx-module-0.32 --add-module=ext/redis2-nginx-module-0.14 --add-module=ext/array-var-nginx-module-0.05 --add-module=ext/nginx-upload-progress-module-0.9.2 --add-module=ext/ngx-fancyindex-0.4.1 --add-module=ext/ngx_metrics-0.1.1 --add-module=ext/nginx_upstream_check_module-10782eaff51872a8f44e65eed89bbe286004bcb1 --add-module=ext/stream-lua-nginx-module-594a297baf73993aa69f64e93f6b623468eb584c --with-http_ssl_module --with-mail --with-mail_ssl_module --with-stream --with-stream_ssl_module --user=nginx --group=nginx`

Although it's built on my fork of the repo https://github.com/dreamcommerce/ngx_http_enhanced_memcached_module  But there are no important changes right now.

**Comment by tiberius** [comment]
Hey, 

thanks for your reply. Even with the fork the error stays the same with 1.10.x, up to 1.11.13. I remember that I had it working successfully, so something must have changed in between ... anyways on some pages the error is subtle, sometimes I actually have to look in the source code of the returned page to "see" that the data has been truncated ... 

I ripped down my own configuration (disabled headers more module as well as google ngx_pagespeed module):

```
root@eb87297081d2:/# nginx -V
nginx version: nginx/1.11.13
built by gcc 4.9.2 (Debian 4.9.2-10) 
built with OpenSSL 1.0.1t  3 May 2016
TLS SNI support enabled
configure arguments: --prefix=/etc/nginx --sbin-path=/usr/sbin/nginx --modules-path=/usr/lib/nginx/modules --conf-path=/etc/nginx/nginx.conf --error-log-path=/var/log/nginx/error.log --http-log-path=/var/log/nginx/access.log --pid-path=/var/run/nginx.pid --lock-path=/var/run/nginx.lock --http-client-body-temp-path=/var/cache/nginx/client_temp --http-proxy-temp-path=/var/cache/nginx/proxy_temp --http-fastcgi-temp-path=/var/cache/nginx/fastcgi_temp --http-uwsgi-temp-path=/var/cache/nginx/uwsgi_temp --http-scgi-temp-path=/var/cache/nginx/scgi_temp --user=nginx --group=nginx --with-compat --with-file-aio --with-threads --with-http_addition_module --with-http_auth_request_module --with-http_dav_module --with-http_flv_module --with-http_gunzip_module --with-http_gzip_static_module --with-http_mp4_module --with-http_random_index_module --with-http_realip_module --add-module=/usr/src/ngx_http_enhanced_memcached_module --with-http_secure_link_module --with-http_slice_module --with-http_ssl_module --with-http_stub_status_module --with-http_sub_module --with-http_v2_module --with-mail --with-mail_ssl_module --with-stream --with-stream_realip_module --with-stream_ssl_module --with-stream_ssl_preread_module --with-cc-opt='-g -O2 -fstack-protector-strong -Wformat -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -fPIC' --with-ld-opt='-Wl,-z,relro -Wl,-z,now -Wl,--as-needed -pie'
```
(For the build I'm using https://hub.docker.com/r/library/nginx/ and do a recompile inside docker using the debian build process ... I can supply you with .deb files if you want it)

Anyways I noticed that on the memcached side the data is complete, here's a snippet:

```
$ socat - UNIX-CONNEcT:lock/memcached.sock
get 23bee4d8bf966ae7f51f2f8ca9a15416
VALUE 23bee4d8bf966ae7f51f2f8ca9a15416 0 887
EXTRACT_HEADERS
Expires: Fri, 14 Apr 2017 09:56:15 GMT
X-Nginx-Status: 200
Vary: Cookie
Content-Type: text/html; charset=utf-8
Content-Length: 674
Cache-Control: max-age=2400
X-Frame-Options: SAMEORIGIN

<!DOCTYPE html><html lang="en"><head>[...]
[...]</aside></body></html>

END
```

Here's a snippet from the debug-log produced with nginx-debug, you can see that the data is truncated from 674 bytes (stored in memcached) to 407 bytes (transferred to the client):

```
[...]
mc-nginx_1     | 2017/04/14 09:36:41 [debug] 16#16: *3 malloc: 000056308460EAD0:8192
mc-nginx_1     | 2017/04/14 09:36:41 [debug] 16#16: *3 recv: eof:0, avail:1
mc-nginx_1     | 2017/04/14 09:36:41 [debug] 16#16: *3 recv: fd:15 940 of 8192
mc-nginx_1     | 2017/04/14 09:36:41 [debug] 16#16: *3 enhanced memcached: response: "VALUE 23bee4d8bf966ae7f51f2f8ca9a15416 0 887"
mc-nginx_1     | 2017/04/14 09:36:41 [debug] 16#16: *3 enhanced memcached: extracting headers from memcached value
mc-nginx_1     | 2017/04/14 09:36:41 [debug] 16#16: *3 enhanced memcached: extracted header: "Expires: Fri, 14 Apr 2017 09:56:15 GMT"
mc-nginx_1     | 2017/04/14 09:36:41 [debug] 16#16: *3 enhanced memcached: extracted header: "Vary: Cookie"
mc-nginx_1     | 2017/04/14 09:36:41 [debug] 16#16: *3 enhanced memcached: extracted header: "Content-Type: text/html; charset=utf-8"
mc-nginx_1     | 2017/04/14 09:36:41 [debug] 16#16: *3 enhanced memcached: extracted header: "Content-Length: 674"
mc-nginx_1     | 2017/04/14 09:36:41 [debug] 16#16: *3 enhanced memcached: extracted header: "Cache-Control: max-age=2400"
mc-nginx_1     | 2017/04/14 09:36:41 [debug] 16#16: *3 enhanced memcached: extracted header: "X-Frame-Options: SAMEORIGIN"
mc-nginx_1     | 2017/04/14 09:36:41 [debug] 16#16: *3 enhanced memached: header done
mc-nginx_1     | 2017/04/14 09:36:41 [debug] 16#16: *3 HTTP/1.1 200 OK
mc-nginx_1     | Server: nginx
mc-nginx_1     | Date: Fri, 14 Apr 2017 09:36:41 GMT
mc-nginx_1     | Content-Type: text/html; charset=utf-8
mc-nginx_1     | Transfer-Encoding: chunked
mc-nginx_1     | Connection: keep-alive
mc-nginx_1     | Vary: Accept-Encoding
mc-nginx_1     | Expires: Fri, 14 Apr 2017 09:56:15 GMT
mc-nginx_1     | Vary: Cookie
mc-nginx_1     | Cache-Control: max-age=2400
mc-nginx_1     | X-Frame-Options: SAMEORIGIN
mc-nginx_1     | Allow: GET, HEAD
mc-nginx_1     | X-Cached-By: EMC
mc-nginx_1     | Content-Encoding: gzip
mc-nginx_1     | 
mc-nginx_1     | 2017/04/14 09:36:41 [debug] 16#16: *3 write new buf t:1 f:0 00005630846278E8, pos 00005630846278E8, size: 358 file: 0, size: 0
mc-nginx_1     | 2017/04/14 09:36:41 [debug] 16#16: *3 http write filter: l:0 f:0 s:358
mc-nginx_1     | 2017/04/14 09:36:41 [debug] 16#16: *3 tcp_nodelay
mc-nginx_1     | 2017/04/14 09:36:41 [debug] 16#16: *3 posix_memalign: 0000563084610AE0:4096 @16
mc-nginx_1     | 2017/04/14 09:36:41 [debug] 16#16: *3 enhanced memcached: filter bytes:681 size:681 length:600 rest:7
mc-nginx_1     | 2017/04/14 09:36:41 [error] 16#16: *3 enhanced memcached: sent invalid trailer while reading response header from upstream, client: 172.17.0.1, server: example.xxx.xx, request: "GET / HTTP/1.1", upstream: "memcached://unix:/var/lock/memcached.sock:", host: "example.xxx.xx"
mc-nginx_1     | 2017/04/14 09:36:41 [debug] 16#16: *3 http upstream process non buffered downstream
mc-nginx_1     | 2017/04/14 09:36:41 [debug] 16#16: *3 http output filter "/?"
mc-nginx_1     | 2017/04/14 09:36:41 [debug] 16#16: *3 http copy filter: "/?"
mc-nginx_1     | 2017/04/14 09:36:41 [debug] 16#16: *3 http postpone filter "/?" 0000563084627A50
mc-nginx_1     | 2017/04/14 09:36:41 [debug] 16#16: *3 http gzip filter
mc-nginx_1     | 2017/04/14 09:36:41 [debug] 16#16: *3 malloc: 000056308461D230:16384
mc-nginx_1     | 2017/04/14 09:36:41 [debug] 16#16: *3 gzip alloc: n:1 s:5936 a:8192 p:000056308461D230
mc-nginx_1     | 2017/04/14 09:36:41 [debug] 16#16: *3 gzip alloc: n:1024 s:2 a:2048 p:000056308461F230
mc-nginx_1     | 172.17.0.1 - - [14/Apr/2017:09:36:41 +0000] "GET / HTTP/1.1" 200 407 "-" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Ubuntu Chromium/57.0.2987.98 Chrome/57.0.2987.98 Safari/537.36"
mc-nginx_1     | 2017/04/14 09:36:41 [debug] 16#16: *3 gzip alloc: n:1024 s:2 a:2048 p:000056308461FA30
mc-nginx_1     | 2017/04/14 09:36:41 [debug] 16#16: *3 gzip alloc: n:1024 s:2 a:2048 p:0000563084620230
mc-nginx_1     | 2017/04/14 09:36:41 [debug] 16#16: *3 gzip alloc: n:512 s:4 a:2048 p:0000563084620A30
mc-nginx_1     | 2017/04/14 09:36:41 [debug] 16#16: *3 gzip in: 0000563084627A70
mc-nginx_1     | 2017/04/14 09:36:41 [debug] 16#16: *3 gzip in_buf:0000563084610B00 ni:000056308460EBD3 ai:593
mc-nginx_1     | 2017/04/14 09:36:41 [debug] 16#16: *3 malloc: 000056308462A020:8192
mc-nginx_1     | 2017/04/14 09:36:41 [debug] 16#16: *3 deflate in: ni:000056308460EBD3 no:000056308462A020 ai:593 ao:8192 fl:2 redo:0
mc-nginx_1     | 2017/04/14 09:36:41 [debug] 16#16: *3 deflate out: ni:000056308460EE24 no:000056308462A192 ai:0 ao:7822 rc:0
mc-nginx_1     | 2017/04/14 09:36:41 [debug] 16#16: *3 gzip in_buf:0000563084610B00 pos:000056308460EBD3
mc-nginx_1     | 2017/04/14 09:36:41 [debug] 16#16: *3 http chunk: 10
mc-nginx_1     | 2017/04/14 09:36:41 [debug] 16#16: *3 http chunk: 370
```

Please let me know if I can assist you in further debugging this problem.

**Comment by manwe-pl** [comment]
Well. Maybe the easiest way would be if I test your config. Anonymize it, strip down to bare minimum and post here. I'll check if I can replicate the same problem. 

I'm using Gentoo, not docker or Debian.

**Comment by tiberius** [comment]
Hey, 

tldr: I have a strong guess, that it's related to my backend already setting a Content-Length header, see below. 

But for the completeness:

/etc/nginx/nginx.conf (mostly typical settings, I guess):
```
user nginx;
worker_processes 8;
pid /run/nginx.pid;
worker_rlimit_nofile 20480;

events {
   worker_connections 1024;
   multi_accept on;
}

http {

   ##
   # Basic Settings
   ##

   sendfile on;
   tcp_nopush on;
   tcp_nodelay on;
   keepalive_timeout 2;
   types_hash_max_size 2048;
   server_tokens off;
   server_names_hash_max_size 1024;
   server_names_hash_bucket_size 128;

   # server_names_hash_bucket_size 64;
   # server_name_in_redirect off;

   include /etc/nginx/mime.types;
   default_type application/octet-stream;

   ##
   # Logging Settings
   ##

   access_log /dev/stdout combined;
   error_log stderr;

   ##
   # File Cache Settings
   ##

   open_file_cache          max=5000  inactive=20s;
   open_file_cache_valid    30s;
   open_file_cache_min_uses 2;
   open_file_cache_errors   on;

   ##
   # Gzip Settings
   ##

   gzip on;
   gzip_static on;
   gzip_disable "msie6";
   gzip_http_version 1.1;
   gzip_vary on;
   gzip_comp_level 6;
   gzip_proxied any;
   gzip_types text/plain text/css application/json application/x-javascript text/xml application/xml application/xml+rss text/javascript application/javascript text/x-js image/svg+xml;
   gzip_buffers 16 8k;

   ssl_session_cache    shared:SSL:10m;
   ssl_session_timeout  10m;
   ssl_ciphers EECDH+AESGCM:EDH+AESGCM:EECDH:EDH:!MD5:!RC4:!LOW:!MEDIUM:!CAMELLIA:!ECDSA:!DES:!DSS:!3DES:!NULL;
   ssl_prefer_server_ciphers on;
   ssl_protocols TLSv1 TLSv1.1 TLSv1.2;
   ssl_dhparam /etc/ssl/private/dhparams.pem;

   ##
   # Virtual Host Configs
   ##

   include /etc/nginx/conf.d/*.conf;
   include /etc/nginx/sites-available/*.conf;
}
```

Minimal site-config, /etc/nginx/sites-available/example.com.conf:
```
upstream frontcache {
	server unix:/var/lock/memcached.sock;
}

server {
	listen 80;
	server_name example.com;

	access_log /dev/stdout combined;
	error_log /dev/stderr debug;

	location / {
		add_header Allow "GET, HEAD" always;

		set $enhanced_memcached_key "$scheme://$host$request_uri";

		enhanced_memcached_connect_timeout 25ms;
		enhanced_memcached_read_timeout 25ms;
		enhanced_memcached_send_timeout 25ms;
		enhanced_memcached_buffer_size 512k;
		enhanced_memcached_hash_keys_with_md5 off;
		enhanced_memcached_pass frontcache;

	}

}
```

I have an entry in my `/etc/hosts` file for that hostname. But that's obviously not the problem anyways.

For simplicity I also ripped out the original backend, so I'm writing the data to memcached by hand:
```
$ printf "set http://example.com/ 0 3600 $(expr $(wc -c /tmp/doc.html | cut -d ' ' -f 1) - 1)\r\n$(cat /tmp/doc.html)\r\n" | socat - UNIX-CONNECT:lock/memcached.sock
STORED
```

contents of /tmp/doc.html:
```
$ cat /tmp/doc.html 
EXTRACT_HEADERS
X-Nginx-Status: 200
Content-Length: 168
Content-Type: text/html; charset=utf-8

<!DOCTYPE html><html lang="en"><head><title>demo</title></head><body><h1>Hello World</h1></body></html>
```

The magic seems to be the Content-Length header; if i'm only counting the real content, the value should be:
```
$ tail -n 1 /tmp/doc.html | wc -c
105
```

But the value given above (168) is calculated by the means of the last four lines (might be coincidental, I found it by try'n'error. 166 works for me, too):
```
$ tail -n 4 /tmp/doc.html | wc -c
168
```

From what I know right now, is that omitting the `Content-Length` header alltogether does the job as expected. If I have to, I'm probably able to ripp that one out of the response of my backend prior to storing. Anyways, if I set the Content-Length in the stored response to the value of 105 it get's truncated to a much shorter value:

```
$ wget -O/tmp/response --server-response http://example.com/; cat /tmp/response
--2017-04-14 18:41:19--  http://example.com/
Resolving example.com (example.com)... 127.0.2.1
Connecting to example.com (example.com)|127.0.2.1|:80... connected.
HTTP request sent, awaiting response... 
  HTTP/1.1 200 OK
  Server: nginx
  Date: Fri, 14 Apr 2017 16:41:19 GMT
  Content-Type: text/html; charset=utf-8
  Content-Length: 42
  Connection: keep-alive
  Vary: Accept-Encoding
  Allow: GET, HEAD
Length: 42 [text/html]
Saving to: ‘/tmp/response’

/tmp/response             100%[===================================>]      42  --.-KB/s    in 0s      

2017-04-14 18:41:19 (4,45 MB/s) - ‘/tmp/response’ saved [42/42]

<!DOCTYPE html><html lang="en"><head><titl
```

I hope I made everything clear. Thanks in advance!

**Comment by bpaquet** [comment]
Why not using the ngnx module to store data in memcache ?

Le ven. 14 avr. 2017 à 10:02, Kai Nessig <notifications@github.com> a
écrit :

**Comment by tiberius** [comment]
Hey, 

well, the way to store the data isn't the problem at all. In general I'm using django as a backend to store data to memcached, it was only for the purpose of simplicity (though it might not be that simple). 

The thing for me is, I nailed it down: once i drop the Content-Length header from the backend response, there's no error, so:

```
EXTRACT_HEADERS
X-Nginx-Status: 200
Content-Type: text/html; charset=utf-8

<!DOCTYPE html><html lang="en">[...]
```

... works as expected (though there is no Content-Length header then at all). 


```
EXTRACT_HEADERS
X-Nginx-Status: 200
Content-Length: <Actual content length + extra>
Content-Type: text/html; charset=utf-8

<!DOCTYPE html><html lang="en">[...]
```

... works as well, but it will be hard to calculate that value reliably.

```
EXTRACT_HEADERS
X-Nginx-Status: 200
Content-Length: <Actual content length of response body>
Content-Type: text/html; charset=utf-8

<!DOCTYPE html><html lang="en">[...]
```

... truncates data and gives an error.

**Comment by bpaquet** [comment]
So why do you want to set Content-Length in EXTRACT_HEADERS ?

Le ven. 14 avr. 2017 à 12:18, Kai Nessig <notifications@github.com> a
écrit :

**Comment by tiberius** [comment]
Well, the django backend does it automatically (I actually have to remove it again from the response), also RFC says it should be set (https://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html Section 14.13), when it's available.

**Comment by bpaquet** [comment]
Nginx will populate the header automatically if not set in EXTRACT_HEADERS.

Le ven. 14 avr. 2017 à 12:53, Kai Nessig <notifications@github.com> a
écrit :

**Comment by tiberius** [comment]
Hey, 

that's sadly not true, when uwsgi_pass is in use, as it seems. Anyways, now that I know about it, I can work-around it. Maybe it's enough if you just add a note about *not* setting Content-Length in EXTRACT_HEADERS. 

Have you played around a bit with the values? Looks like there's something overflowing if the value is too large (one can actually get the END token from memcached get printed on the page), e. g. if you set the value to 200 bytes in the example above).

**Comment by bpaquet** [comment]
With *pass, nginx will (most of the time) use a chunked transfer, and will
not set Content-Length. If you have an easy way to reproduce the bug, I
will be happy to have a look.

Le sam. 15 avr. 2017 à 02:22, Kai Nessig <notifications@github.com> a
écrit :

**Comment by tiberius** [comment]
Well as shown here (https://github.com/bpaquet/ngx_http_enhanced_memcached_module/issues/5#issuecomment-294218207), to reproduce the bug, you basically just need to store the Content-Length header within EXTRACT_HEADERS

**Comment by bpaquet** [comment]
Fixed. Thx for reporting.

Le sam. 22 avr. 2017 à 08:58, Kai Nessig <notifications@github.com> a
écrit :
